File: docs/Drawable/LogoModule/Turtle.js
import { COLOR } from "../../System/Constants.js";
import { Line, Flood } from "../RenderItems.js";

export class Turtle {
    constructor() {
        [this.pos_x, this.pos_y] = [315, 300];
        this.base = {
            "x": this.pos_x,
            "y": this.pos_y
        };
        this.rotation = 0;
        this.show_turtle = true;
        this.pen_color = COLOR.BLACK;
        this.pen_down = true;
        this.pen_width = 1;
        this.distortion = 0;

        this.lines = [];

        this.boundaries = {
            "top": null,
            "bot": null
        };
    }

    Forward(distance = 20) {
        var [start_x, start_y] = [this.pos_x, this.pos_y];
        [this.pos_x, this.pos_y] = this.Move(1, distance);
        this.NormalizeMovement();
        var [end_x, end_y] = [this.pos_x, this.pos_y];
        var line_o = new Line([start_x + this.distortion, start_y + this.distortion], [end_x + this.distortion, end_y + this.distortion],
            this.pen_color, this.pen_width, this.pen_down);
        this.lines.push(line_o);
    }

    Backward(distance = 20) {
        var [start_x, start_y] = [this.pos_x, this.pos_y];
        [this.pos_x, this.pos_y] = this.Move(-1, distance);
        this.NormalizeMovement();
        var [end_x, end_y] = [this.pos_x, this.pos_y];
        var line_o = new Line([start_x + this.distortion, start_y + this.distortion], [end_x + this.distortion, end_y + this.distortion],
            this.pen_color, this.pen_width, this.pen_down);
        this.lines.push(line_o);
    }

    Right(angle = 90) {
        this.rotation += angle;
        this.rotation %= 360;
    }

    Left(angle = 90) {
        this.rotation -= angle;
        this.rotation %= 360;
    }

    Home() {
        var [start_x, start_y] = [this.pos_x, this.pos_y];
        [this.pos_x, this.pos_y] = [this.base.x, this.base.y];
        var [end_x, end_y] = [this.pos_x, this.pos_y];
        var line_o = new Line([start_x + this.distortion, start_y + this.distortion], [end_x + this.distortion, end_y + this.distortion],
            this.pen_color, this.pen_width, this.pen_down);
        this.lines.push(line_o);
    }

    PenDown() {
        this.pen_down = true;
    }

    PenUp() {
        this.pen_down = false;
    }

    Width(size) {
        this.pen_width = size;
    }

    Color(color) {
        this.pen_color = color;
    }

    Fill() {
        var flood = new Flood([this.pos_x, this.pos_y], this.pen_color);
        this.lines.push(flood);
    }

    Reset() {
        [this.pos_x, this.pos_y] = [this.base.x, this.base.y];
        this.lines = [];
        this.rotation = 0;
        this.pen_color = COLOR.BLACK;
        this.pen_width = 3;
        this.pen_down = true;
        this.show_turtle = true;
    }

    Clear() {
        this.lines = []
    }

    ShowTurtle() {
        this.show_turtle = true;
    }

    HideTurtle() {
        this.show_turtle = false;
    }

    Move(direction, distance) {
        return [Math.trunc(Math.round(this.pos_x + (distance * Math.cos(Math.PI / 180 * this.rotation)) * direction)),
        Math.trunc(Math.round(this.pos_y + (distance * Math.sin(Math.PI / 180 * this.rotation)) * direction))];
    }

    Debug() {
        console.log(this.pos_x, this.pos_y);
        console.log(this.rotation);
    }


    GetTurtleInformationToRender() {
        return [[this.pos_x, this.pos_y], this.rotation, this.show_turtle]
    }

    GetLinesForRenderer() {
        return this.lines
    }

    SetDistortion(amount) {
        this.distortion = amount;
    }

    SetBoundaries(top, bot) {
        this.boundaries.top = top;
        this.boundaries.bot = bot;
    }

    NormalizeMovement() {
        if (this.boundaries.top != null && this.boundaries["bot"] != null) {
            if (this.pos_x < this.boundaries.top[0]) {
                this.pos_x = this.boundaries.top[0];
            }
            if (this.pos_x > this.boundaries.bot[0]) {
                this.pos_x = this.boundaries.bot[0];
            }
            if (this.pos_y < this.boundaries.top[1]) {
                this.pos_y = this.boundaries.top[1];
            }
            if (this.pos_y > this.boundaries.bot[1]) {
                this.pos_y = this.boundaries.bot[1];
            }

        }
    }

}

File: szakdoliV3/Drawable/LogoModule/Turtle.py
# -*- coding: utf-8 -*-

from math import sin, cos, radians

from Drawable.RenderItems import Line, Flood
from System.Constants import COLOR as Color


class Turtle:

    def __init__(self):
        self.pos_x , self.pos_y = (315, 300)
        self.base = {
            "x" : self.pos_x,
            "y" : self.pos_y
        }
        self.rotation = 0
        self.show_turtle = True
        self.pen_color = Color.BLACK
        self.pen_down = True
        self.pen_witdh = 1
        self.distorsion = 0

        self.lines = [] # <instance Class : Line >

        self.boundaries = {
            "top" : None,
            "bot" : None
        }


    def forward(self, distance = 20):
        start_x, start_y = self.pos_x, self.pos_y
        self.pos_x, self.pos_y = self.move(1, distance)
        self.NormalizeMovement()
        end_x, end_y = self.pos_x, self.pos_y
        line_o = Line((start_x + self.distorsion, start_y + self.distorsion), (end_x + self.distorsion, end_y + self.distorsion),
                      self.pen_color, self.pen_witdh, self.pen_down)
        self.lines.append(line_o)

    def backward(self, distance = 20):
        start_x, start_y = self.pos_x, self.pos_y
        self.pos_x, self.pos_y = self.move(-1, distance)
        self.NormalizeMovement()
        end_x, end_y = self.pos_x, self.pos_y
        line_o = Line((start_x, start_y), (end_x, end_y), self.pen_color, self.pen_witdh, self.pen_down)
        self.lines.append(line_o)

    def right(self, angle = 90):
        self.rotation += angle
        self.rotation %= 360

    def left(self, angle = 90):
        self.rotation -= angle
        self.rotation %= 360

    def home(self):
        start_x, start_y = self.pos_x, self.pos_y
        self.pos_x, self.pos_y = self.base["x"], self.base["y"]
        end_x, end_y = self.pos_x, self.pos_y
        line_o = Line((start_x + self.distorsion, start_y + self.distorsion), (end_x + self.distorsion, end_y + self.distorsion),
                      self.pen_color, self.pen_witdh, self.pen_down)
        self.lines.append(line_o)

    def pendown(self):
        self.pen_down = True

    def penup(self):
        self.pen_down = False

    def width(self, size):
        self.pen_witdh = size

    def color(self, color):
        self.pen_color = color

    def fill(self):
        flood = Flood((self.pos_x, self.pos_y), self.pen_color)
        self.lines.append(flood)

    def reset(self):
        self.pos_x, self.pos_y = self.base["x"], self.base["y"]
        self.lines = []
        self.rotation = 0
        self.pen_color = Color.BLACK
        self.pen_witdh = 3
        self.pen_down = True
        self.show_turtle = True

    def clear(self):
        self.lines = []

    def showturtle(self):
        self.show_turtle = True

    def hideturtle(self):
        self.show_turtle = False


    def move(self, direction, distance):
        return int(round(self.pos_x + (distance * cos(radians(self.rotation))) * direction)),\
               int(round(self.pos_y + (distance * sin(radians(self.rotation))) * direction))

    def debug(self):
        print (self.pos_x, self.pos_y)
        print (self.rotation)

    def GetTurtleInformationToRender(self):
        return (self.pos_x, self.pos_y), self.rotation, self.show_turtle

    def GetLinesForRenderer(self):
        return self.lines

    def SetDistorsion(self, amount):
        self.distorsion = amount

    def SetBoundaries(self, top, bot):
        self.boundaries["top"] = top
        self.boundaries["bot"] = bot

    def NormalizeMovement(self):
        if self.boundaries["top"] is not None and self.boundaries["bot"] is not None:

            if self.pos_x < self.boundaries["top"][0]:
                self.pos_x = self.boundaries["top"][0]
            if self.pos_x > self.boundaries["bot"][0]:
                self.pos_x = self.boundaries["bot"][0]

            if self.pos_y < self.boundaries["top"][1]:
                self.pos_y = self.boundaries["top"][1]
            if self.pos_y > self.boundaries["bot"][1]:
                self.pos_y = self.boundaries["bot"][1]